README.md
# NAME

Catmandu::EXL - Tools for ALEPHXML and PNX records

# SYNOPSIS

    # On the command line

    catmandu convert MARC --type ALEPHXML to YAML < ./t/aleph.xml
    catmandu convert MARC --type ALEPHXML --fix ./t/pnx.fix to PNX < ./t/aleph.xml

    # In Perl 

    use EXL::Parser::ALEPHXML;

    my $parser = EXL::Parser::ALEPHXML->new( './t/aleph.xml' );
    my $record = $parser->next();

    # In Perl with Catmandu

    use Catmandu;
    use Catmandu::Importer::MARC;

    my $importer = Catmandu::Importer::MARC->new(file => "./t/aleph.xml", type => "ALEPHXML");
    my $records = $importer->to_array();

# DESCRIPTION

_Experimental release_

Catmandu::EXL  provides methods, classes, and functions to process ALEPHXML and PNX records in Perl.

ALEPHXML is an export data format of Aleph library systems used in Germany. It is a mix of the Machine Readable Cataloging format (MARC) and the Maschinelles Austauschformat fuer Bibliotheken (MAB), see [https://wiki1.hbz-nrw.de/pages/viewpage.action?pageId=13238276](https://wiki1.hbz-nrw.de/pages/viewpage.action?pageId=13238276) for a more detailed description.

Primo Normalized XML (PNX) is a data format for the Primo discovery system.  

Records in Catmandu::EXL are encoded either as as array of arrays, the inner arrays representing record fields, or as an object with two fields, \_id and record, the latter holding the record as array of arrays, and the former holding the record identifier, stored in field 001.

# AUTHOR

Johann Rolschewski <rolschewski@gmail.com>

# COPYRIGHT

Copyright 2015- Johann Rolschewski

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO

[Catmandu](https://metacpan.org/pod/Catmandu).


Changes
Revision history for Catmandu-EXL

{{$NEXT}}

0.02  2015-01-17 17:06:34 CET
        - add params to PNX exporter
        - add fix

0.01  2015-01-14 17:58:28 CET
        - Initial release


lib/Catmandu/Fix/lds38.pm
package Catmandu::Fix::lds38;
use Catmandu::Sane;
use Catmandu::Util qw(is_string);

sub new {
    my ($class,$path) = @_;
    return bless { path => $path } , $class;
}

sub fix {
    my ($self,$data) = @_;
    my @path  = split(/\./,$self->{path});
    my $key = eval '$data'.(join '', map "->{$_}", @path ) or die "path ". $self->{path} . " does not exists";
    $data->{pnx}->{display}->{lds38} = [] unless exists $data->{pnx}->{display}->{lds38};
    foreach my $arr_ref (@{$key}) {
        if (is_string($arr_ref->[0]) && $arr_ref->[0] eq 'SRD') {
            my $lds38 = $arr_ref->[1] . ' FUB_ALEPH' . $arr_ref->[2];
            $lds38 =~ s/\^/ /g;
            push @{$data->{pnx}->{display}->{lds38}}, $lds38;
         } 
    }
    $data;
}

=head1 NAME
 
Catmandu::Fix::lds38 - custom Fix for PNX element <lds38>
 
=head1 SYNOPSIS
 
    # get subfields in sorted order, join them and store them in tmp variable
    marc_map('PLKanb','tmp.plk.$append', -pluck => 1, -join => '~~~')
    # split string to array
    split_field('tmp.plk.*','~~~')
    # apply fix on given path
    lds38('tmp.plk')
    remove_field('tmp.plk')

=head1 DESCRPTION

The datastructure referenced by 'path' must be an array of arrays ([[$sf_a, $sf_n, $sf_b],...])

Checks if subfield 'a' matches 'SDR' and adds new element to path 'pnx.display.lds38'.

=head1 SEE ALSO
 
L<Catmandu::Fix>, L<Catmandu::EXL>.
 
=cut

1;

lib/EXL/Parser/ALEPHXML.pm
package EXL::Parser::ALEPHXML;

our $VERSION = '0.02';

use strict;
use warnings;
use Carp qw<croak>;
use XML::LibXML::Reader;


sub new {
    my $class = shift;
    my $input  = shift;

    my $self = {
        filename   => undef,
        rec_number => 0,
        xml_reader => undef,
    };

    # check for file or filehandle
    my $ishandle = eval { fileno($input); };
    if ( !$@ && defined $ishandle ) {
        binmode $input; # drop all PerlIO layers, as required by libxml2
        my $reader = XML::LibXML::Reader->new( IO => $input )
            or croak "cannot read from filehandle $input\n";
        $self->{filename}   = scalar $input;
        $self->{xml_reader} = $reader;
    }
    elsif ( defined $input && $input !~ /\n/ && -e $input ) {
        my $reader = XML::LibXML::Reader->new( location => $input )
            or croak "cannot read from file $input\n";
        $self->{filename}   = $input;
        $self->{xml_reader} = $reader;
    }
    elsif ( defined $input && length $input > 0 ) {
        my $reader = XML::LibXML::Reader->new( string => $input )
            or croak "cannot read XML string $input\n";
        $self->{xml_reader} = $reader;
    }
    else {
        croak "file, filehande or string $input does not exists";
    }
    return ( bless $self, $class );
}


sub next {
    my $self = shift;

    # get only <record> elements from 'mabxml' namespace
    # skip <record> elements from 'oai' namespace
    if ( $self->{xml_reader}->nextElement('record', 'http://www.ddb.de/professionell/mabxml/mabxml-1.xsd') ) {
        $self->{rec_number}++;
        my $record = _decode( $self->{xml_reader} );
        return $record;
    }
    return;
}


sub _decode {
    my $reader = shift;
    my @record;

    # get all field nodes from record;
    foreach my $field_node (
        $reader->copyCurrentNode(1)->getChildrenByTagName('*') )
    {
        my @field;

        # get field tag number or skip element (e.g. <leader>)
        my $tag = $field_node->getAttribute('tag') or next;

        # get 1. indicator, ignore second
        my $ind1 = $field_node->getAttribute('ind1') // '';
        $ind1 =~ s/-/ /;
        my $ind2 = $field_node->getAttribute('ind2') // '';
        $ind2 =~ s/-/ /;

        # ToDo: textContent ignores </tf> and <ns>

        # Check for data or subfields
        if ( my @subfields = $field_node->getChildrenByTagName('*') ) {
            push( @field, ( $tag, $ind1, $ind2 ) );

            # get all subfield nodes
            foreach my $subfield_node (@subfields) {
                my $subfield_code = $subfield_node->getAttribute('code');
                my $subfield_data = $subfield_node->textContent;
                push( @field, ( $subfield_code, $subfield_data ) );
            }
        }
        else {
            my $data = $field_node->textContent();
            push( @field, ( $tag, $ind1, $ind2, '_', $data ) );
        }

        push( @record, [@field] );
    }
    return \@record;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

EXL::Parser::ALEPHXML - ALEPHXML parser

=head1 SYNOPSIS

EXL::Parser::ALEPHXML is a parser for ALEPHXML records.

    use EXL::Parser::ALEPHXML;

    my $parser = EXL::Parser::ALEPHXML->new( $filename );

    while ( my $record_hash = $parser->next() ) {
        # do something        
    }

=head1 Arguments

=over

=item C<file>

Path to file with ALEPHXML records.

=item C<fh>

Open filehandle for file with ALEPHXML records.

=item C<string>

XML string with ALEPHXML records.

=back

=head1 METHODS

=head2 new($filename | $filehandle | $string)

=head2 next()

Reads the next record from ALEPHXML input stream. Returns a Perl hash.

=head2 _decode($record)

Deserialize a ALEPHXML record to an an ARRAY of ARRAYs.

=head1 SEEALSO

L<Catmandu::EXL>.

=head1 AUTHOR

Johann Rolschewski <rolschewski@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Johann Rolschewski.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Catmandu/Exporter/EXL/PNX.pm
package Catmandu::Exporter::EXL::PNX;
use Catmandu::Sane;
use Catmandu::Util qw(xml_escape is_different :array :is);
use Moo;

with 'Catmandu::Exporter';

has record          => ( is => 'ro', default => sub {'record'} );
has xml_declaration => ( is => 'ro', default => sub { 0 } );
has collection      => ( is => 'ro' , default => sub { 0 } );
has _n              => ( is => 'rw', default => sub { 0 } );

# ToDo:
# https://github.com/dtulibrary/finditproxy/blob/master/xsd/primo_nm_bib.xsd
# set namespace
# use LibXML
# define root element

sub add {
    my ( $self, $data ) = @_;

    my $xml = "";

    if ( $self->_n == 0 ) {
        if ( $self->xml_declaration ) {
            $xml .= Catmandu::Util::xml_declaration;
        }
        if ($self->collection) {
            $xml .= '<PrimoNMBib>';
        }
        $self->_n(1);
    }

    $xml .= '<record>';

    my $record = $data->{pnx};

    foreach my $main_entry ( sort keys %{$record} ) {
        $xml .= "<$main_entry>";
        foreach my $sub_entry ( sort keys %{ $record->{$main_entry} } ) {
            my $content = $record->{$main_entry}->{$sub_entry};
            if ( ref $content eq 'ARRAY' ) {
                $xml .= "<$sub_entry>" . xml_escape($_) . "</$sub_entry>" for @{$content};

            }
            else {
                $xml .= "<$sub_entry>" . xml_escape($content) . "</$sub_entry>";
            }
        }
        $xml .= "</$main_entry>";
    }

    $xml .= '</record>';
    $self->fh->print($xml);
}

sub commit {
    my ($self) = @_;

    if($self->collection){
        $self->fh->print('</PrimoNMBib>');
    }

    $self->fh->flush;
}


1;

__END__

=encoding UTF-8

=head1 NAME

Catmandu::Exporter::EXL::PNX - Exporter for PNX records

=head1 SYNOPSIS

    # From the command line 
    $ catmandu convert MARC --type ALEPHXML --fix ./t/pnx.fix to EXL --type PNX < ./t/aleph.xml
    $ catmandu convert MARC --type ALEPHXML --fix ./t/pnx.fix to EXL --type PNX --xml_declaration 1 < ./t/aleph.xml

    # From Perl
    use Catmandu;

    my $importer = Catmandu::Importer::MARC( file => "./t/aleph.xml" , type => 'ALEPHXML');
    my $fixer    = Catmandu->fixer("./t/pnx.fix");
    my $exporter = Catmandu::Exporter::EXL->new( file => "pnx.xml", type => "PNX" );

    $exporter->add_many($fixer->fix($importer));
    $exporter->commit;

=head1 METHODS

=head2 new(file => $file , %opts)

Create a new L<Catmandu::Exporter::EXL::PNX> to serialize PNX records into XML. Provide the path
of a $file to write exported records to. Optionally the following paramters can be
specified:

    xml_declaration : add a xml declaration when true (default: true)

=head1 INHERTED METHODS

=head2 count

=head2 add($hashref)

=head2 add_many($array)

=head2 add_many($iterator)

=head2 add_many(sub {})

=head2 ...

All the L<Catmandu::Exporter> methods are inherited.

=head1 SEE ALSO

L<Catmandu::Exporter>

=cut


lib/Catmandu/EXL.pm
package Catmandu::EXL;

use strict;
use 5.008_005;
our $VERSION = '0.02';

1;
__END__

=encoding UTF-8

=head1 NAME

Catmandu::EXL - Tools for ALEPHXML and PNX records

=head1 SYNOPSIS

    # On the command line

    catmandu convert MARC --type ALEPHXML to YAML < ./t/aleph.xml
    catmandu convert MARC --type ALEPHXML --fix ./t/pnx.fix to PNX < ./t/aleph.xml

    # In Perl 

    use EXL::Parser::ALEPHXML;

    my $parser = EXL::Parser::ALEPHXML->new( './t/aleph.xml' );
    my $record = $parser->next();

    # In Perl with Catmandu

    use Catmandu;
    use Catmandu::Importer::MARC;

    my $importer = Catmandu::Importer::MARC->new(file => "./t/aleph.xml", type => "ALEPHXML");
    my $records = $importer->to_array();

=head1 DESCRIPTION

I<Experimental release>

Catmandu::EXL  provides methods, classes, and functions to process ALEPHXML and PNX records in Perl.

ALEPHXML is an export data format of Aleph library systems used in Germany. It is a mix of the Machine Readable Cataloging format (MARC) and the Maschinelles Austauschformat fuer Bibliotheken (MAB), see L<https://wiki1.hbz-nrw.de/pages/viewpage.action?pageId=13238276> for a more detailed description.

Primo Normalized XML (PNX) is a data format for the Primo discovery system.  

Records in Catmandu::EXL are encoded either as as array of arrays, the inner arrays representing record fields, or as an object with two fields, _id and record, the latter holding the record as array of arrays, and the former holding the record identifier, stored in field 001.

=head1 AUTHOR

Johann Rolschewski E<lt>rolschewski@gmail.comE<gt>

=head1 COPYRIGHT

Copyright 2015- Johann Rolschewski

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<Catmandu>.

=cut
